File: modules/Bare.f90
Module Bare
  use Common
  implicit None

  private
! public: G0

! interface G0
!    module procedure &
  public :: &
    FLatFreq, &
    FLocFreq, &
    FTau, &
    BTau, &
    FFreq, &
    BLocFreq, &
    BLatFreq, &
    FLocTau, &
    FLatTau, &
    BLocTau, &
    BLatTau, &
    BFreq

! end interface Dyson

contains

  subroutine FLatFreq(norb,ns,nk,nfreq,hlatt,freq,glatt)
    implicit none

    integer, intent(in) :: norb,ns,nk,nfreq
    double precision, intent(in) :: freq(0:(nfreq-1))
    complex*16, intent(in) :: hlatt(norb,norb,ns,nk)
    complex*16,intent(out) :: glatt(norb,norb,ns,nk,0:(nfreq-1))

    integer :: ik

    glatt=0.0d0

    do ik=1, nk
      call FLocFreq(norb,ns,nfreq,hlatt(:,:,:,ik), freq,glatt(:,:,:,ik,:))
    enddo

  end subroutine FLatFreq

  subroutine FLocFreq(norb,ns,nfreq,hloc,freq,gloc)
    implicit none

    integer, intent(in) :: norb,ns,nfreq
    double precision, intent(in) :: freq(0:(nfreq-1))
    complex*16, intent(in) :: hloc(norb,norb,ns)
    complex*16,intent(out) :: gloc(norb,norb,ns,0:(nfreq-1))

    integer :: is,ifreq,iorb,jorb
    double precision :: w(norb)
    complex*16 :: tempmat(norb,norb), ai, gfreq(0:(nfreq-1),norb),tempmat2(norb,norb)

    ai=dcmplx(0.0d0, 1.0d0)
    gloc=0.0d0

    do is=1, ns
      tempmat=hloc(:,:,is)
      call hermitianeigen_dcmplx(norb, w, tempmat)

      gfreq=0.0d0
      do iorb=1, norb
        call FFreq(nfreq, freq, w(iorb), gfreq(:,iorb))
      enddo

      do ifreq=0, nfreq-1
        do iorb=1, norb
          do jorb=1, norb
            tempmat2(iorb,jorb)=tempmat(iorb, jorb)*gfreq(ifreq, jorb)
          enddo
        enddo

        call zgemm('n','c',norb,norb,norb,(1.0d0,0.0d0),tempmat2,norb,tempmat,norb,(0.0d0,0.0d0),gloc(1,1,is,ifreq),norb)
      enddo
    enddo

  end subroutine FLocFreq


  subroutine FTau(ntau, tau, energy, gtau)
! from wikipedia: https://en.wikipedia.org/wiki/Matsubara_frequency
    implicit none
    
    integer, intent(in) :: ntau
    double precision, intent(in) :: tau(0:(ntau-1)), energy
    complex*16, intent(out) :: gtau(0:(ntau-1))    
    
    integer :: unitnum
    double precision :: machep,taumod,taunew, beta, pi
    integer :: itau
    
    pi = datan2(1.0d0, 1.0d0)*4.0d0
    
    beta= tau(0)/(dcos(pi*(ntau-0.5d0)/dble(ntau))+1.0d0)*2.0d0   
    machep = epsilon ( machep )

    do itau=0, ntau-1
      taumod=modulo(tau(itau), beta)
      unitnum=nint(tau(itau)-taumod)/beta
      if (taumod .lt. machep) then
        unitnum=unitnum-1
      endif
      taunew=tau(itau)-beta*unitnum          
 
      
      if (energy .gt. 0) then
        gtau(itau)=(-1)**(unitnum+1)*dexp(-energy*taunew)*(1-1.0d0/(dexp(energy*beta)+1))
      else
        gtau(itau)=(-1)**(unitnum+1)*dexp(-energy*(taunew-beta))*(1.0d0/(dexp(energy*beta)+1))
      endif
    enddo
  end subroutine FTau


  subroutine BTau(ntau, tau, energy, wtau)
! from wikipedia: https://en.wikipedia.org/wiki/Matsubara_frequency
    implicit none
    
    integer, intent(in) :: ntau
    double precision, intent(in) :: tau(0:(ntau-1)), energy
    complex*16, intent(out) :: wtau(0:(ntau-1))    
    
    integer :: unitnum
    double precision :: machep,taumod,taunew, beta, pi
    
    integer :: itau
    
    pi = datan2(1.0d0, 1.0d0)*4.0d0
    
    beta= tau(0)*2.0d0/(dcos(pi*(ntau-0.5d0)/dble(ntau))+1.0d0)    
    
    machep = epsilon ( machep )
    if (dabs(energy) .lt. 1.0d-12) then
      print *, 'zero energy in Bare.BTau. impossible'
      stop
    endif
    
    do itau=0, ntau-1
      taumod=modulo(tau(itau), beta)
      unitnum=nint(tau(itau)-taumod)/beta
      if (taumod .lt. machep) then
        unitnum=unitnum-1
      endif
      taunew=tau(itau)-beta*unitnum          
      
      if (energy .gt. 0) then
        wtau(itau)=-dexp(-energy*taunew)*(1-1.0d0/(dexp(energy*beta)-1))
      else
        wtau(itau)=-dexp(-energy*(taunew-beta))*(1.0d0/(dexp(energy*beta)-1))
      endif
    enddo
  end subroutine BTau
  
  subroutine FFreq(nfreq, freq, energy, gfreq)
! from wikipedia: https://en.wikipedia.org/wiki/Matsubara_frequency
    implicit none
    
    integer, intent(in) :: nfreq
    double precision, intent(in) :: freq(0:(nfreq-1)), energy
    complex*16, intent(out) :: gfreq(0:(nfreq-1))
    
    integer :: ifreq
    complex*16 :: ai
    
    ai=dcmplx(0.0d0, 1.0d0)    
    
    do ifreq=0, nfreq-1
      gfreq(ifreq)=1.0d0/(ai*freq(ifreq)-energy)
    enddo
  end subroutine FFreq


  subroutine BFreq(nfreq, freq, energy, wfreq)
! from wikipedia: https://en.wikipedia.org/wiki/Matsubara_frequency
    implicit none
    
    integer, intent(in) :: nfreq
    double precision, intent(in) :: freq(0:(nfreq-1)), energy
    complex*16, intent(out) :: wfreq(0:(nfreq-1))
    
    
    integer :: ifreq
    complex*16 :: ai
    
    ai=dcmplx(0.0d0, 1.0d0)    
    
    do ifreq=0, nfreq-1
      wfreq(ifreq)=1.0d0/(ai*freq(ifreq)-energy)
    enddo
  end subroutine BFreq  

  subroutine BLocFreq(norb,ns,nfreq,hloc,freq,wloc)
    implicit none

    integer, intent(in) :: norb,ns,nfreq
    double precision, intent(in) :: freq(0:(nfreq-1))
    complex*16, intent(in) :: hloc(norb,norb,ns,ns)
    complex*16,intent(out) :: wloc(norb,norb,ns,ns,0:(nfreq-1))

    integer :: is,js,ifreq,iorb,jorb
    double precision :: w(norb)
    complex*16 :: tempmat(norb,norb), ai, wfreq(0:(nfreq-1),norb),tempmat2(norb,norb)

    ai=dcmplx(0.0d0, 1.0d0)
    wloc=0.0d0

    do is=1, ns
        do js = 1, ns
           tempmat=hloc(:,:,is,js)
           call hermitianeigen_dcmplx(norb, w, tempmat)

           wfreq=0.0d0
           do iorb=1, norb
              call BFreq(nfreq, freq, w(iorb), wfreq(:,iorb))
           enddo

           do ifreq=0, nfreq-1
              do iorb=1, norb
                 do jorb=1, norb
                    tempmat2(iorb,jorb)=tempmat(iorb, jorb)*wfreq(ifreq, jorb)
                  enddo
              enddo

              call zgemm('n','c',norb,norb,norb,(1.0d0,0.0d0),tempmat2,norb,tempmat,norb,(0.0d0,0.0d0),wloc(1,1,is,js,ifreq),norb)

           enddo
        enddo
    enddo
  end subroutine BLocFreq
  
  subroutine BLatFreq(norb,ns,nk,nfreq,hlatt,freq,wlatt)
    implicit none

    integer, intent(in) :: norb, ns, nk, nfreq
    double precision, intent(in) :: freq(0:(nfreq-1))
    complex*16, intent(in) :: hlatt(norb, norb, ns, ns, nk)
    complex*16, intent(out) :: wlatt(norb, norb, ns, ns, nk, 0:(nfreq-1))

    integer :: ik

    wlatt = 0.0d0

    do ik = 1, nk
      call BLocFreq(norb, ns, nfreq, hlatt(:, :, :, :, ik), freq, wlatt(:, :, :, :, ik,:))
    enddo

  end subroutine BLatFreq

  subroutine FLocTau(norb,ns,ntau,hloc,tau,gloc)
    implicit none

    integer, intent(in) :: norb, ns, ntau
    double precision, intent(in) :: tau(0:(ntau-1))
    complex*16, intent(in) :: hloc(norb, norb, ns)
    complex*16, intent(out) :: gloc(norb, norb, ns, 0:(ntau-1))

    integer :: is, itau, iorb, jorb
    double precision :: w(norb)
    complex*16 :: tempmat(norb, norb), ai, gtau(0:(ntau-1),norb), tempmat2(norb,norb)

    ai = dcmplx(0.0d0, 1.0d0)
    gloc = 0.0d0

    do is = 1, ns
      tempmat = hloc(:,:,is)
      call hermitianeigen_dcmplx(norb, w, tempmat)

      gtau = 0.0d0
      do iorb = 1, norb
        call FTau(ntau, tau, w(iorb), gtau(:,iorb))
      enddo

      do itau = 0, ntau-1
        do iorb = 1, norb
          do jorb = 1, norb
            tempmat2(iorb, jorb) = tempmat(iorb, jorb)*gtau(itau, jorb)
          enddo
        enddo

        call zgemm('n', 'c', norb, norb, norb, (1.0d0, 0.0d0), tempmat2, norb,tempmat, norb, (0.0d0, 0.0d0), gloc(1,1,is,itau),norb)
        
      enddo
    enddo
  end subroutine FLocTau 


  subroutine FLatTau(norb,ns,nk,ntau,hlatt,tau,glatt)
    implicit none

    integer, intent(in) :: norb,ns,nk,ntau
    double precision, intent(in) :: tau(0:(ntau-1))
    complex*16, intent(in) :: hlatt(norb,norb,ns,nk)
    complex*16,intent(out) :: glatt(norb,norb,ns,nk,0:(ntau-1))

    integer :: ik

    glatt=0.0d0

    do ik=1, nk
      call FLocTau(norb,ns,ntau,hlatt(:,:,:,ik), tau,glatt(:,:,:,ik,:))
    enddo

  end subroutine FLatTau 

  subroutine BLocTau(norb, ns, ntau, hloc, tau, wloc)
    implicit none
    
    integer, intent(in) :: norb, ns, ntau
    double precision, intent(in) :: tau(0:(ntau-1))
    complex*16, intent(in) :: hloc(norb, norb, ns, ns)
    complex*16, intent(out) :: wloc(norb, norb, ns, ns,0:(ntau-1))
  
    integer :: is, js, itau, iorb, jorb
    double precision :: w(norb)
    complex*16 :: tempmat(norb, norb), ai, wtau(0:(ntau-1), norb), tempmat2(norb,norb)
  
    ai = dcmplx(0.0d0, 1.0d0)
    wloc = 0.0d0
  
    do is = 1, ns
      do js = 1, ns
        tempmat = hloc(:, :, is, js)
        call hermitianeigen_dcmplx(norb, w, tempmat)
  
        wtau = 0.0d0
        do iorb = 1, norb
          call BTau(ntau, tau, w(iorb), wtau(:,iorb))
        enddo
  
        do itau = 0, ntau-1
          do iorb = 1,norb
            do jorb = 1, norb
              tempmat2(iorb, jorb) = tempmat(iorb, jorb)*wtau(itau, jorb)
            enddo
          enddo
  
          call zgemm('n', 'c', norb, norb, norb, (1.0d0, 0.0d0), tempmat2, norb,tempmat, norb, (0.0d0,0.0d0), wloc(1,1, is, js, itau), norb)
  
        enddo
      enddo
    enddo
  
  end subroutine BLocTau

  subroutine BLatTau(norb, ns, nk, ntau, hlatt, tau, wlatt)
    implicit none

    integer, intent(in) :: norb, ns, nk, ntau
    double precision, intent(in) :: tau(0:(ntau-1))
    complex*16, intent(in) :: hlatt(norb, norb, ns, ns, nk)
    complex*16, intent(out) :: wlatt(norb, norb, ns, ns, nk, 0:(ntau-1))

    integer :: ik

    wlatt = 0.0d0

    do ik = 1, nk
      call BLocTau(norb, ns, ntau, hlatt(:,:,:,:,ik), tau, wlatt(:,:,:,:,ik,:))
    enddo

  end subroutine BLatTau

 
end Module Bare

File: src/QAssemble/Src_mpi/modules/Bare.py
from Common import Common
from sys import exit
import numpy as np

class Bare:

    @staticmethod
    def FFreq(freq : np.ndarray, energy : np.ndarray) -> np.ndarray:

        nfreq = len(freq)
        gfreq = np.ndarray((nfreq), dtype=np.ndarray, order='F')

        for ifreq in range(nfreq):
            gfreq[ifreq] = 1.0/(1j*freq[ifreq] - energy)

        return gfreq
    
    @staticmethod
    def FTau(tau : np.ndarray, energy : np.ndarray) -> np.ndarray:

        ntau = len(tau)
        gtau = np.ndarray((ntau), dtype=np.complex128, order='F')

        pi = np.pi
        beta = tau[0]/(np.cos(pi*(ntau-0.5)/ntau) + 1.0)*2.0
        machep = np.finfo(np.float64).eps
        
        for itau in range(ntau):
            taumod = (tau[itau] % beta)
            unitnum = int(tau[itau] - taumod)/beta
            
            if (taumod < machep):
                unitnum = unitnum-1

            taunew = tau[itau] - beta * unitnum
            
            if (energy > 0):
                gtau[itau] = (-1)**(unitnum+1)*np.exp(-energy*taunew, dtype=np.complex128) \
                    * (1 - 1/(np.exp(energy*beta, dtype=np.complex128) + 1))
            else:
                gtau[itau] = (-1)**(unitnum+1)* np.exp(-energy*(taunew-beta), dtype=np.complex128)\
                    * (1.0/(np.exp(energy*beta, dtype=np.complex128) + 1))

        return gtau

    @staticmethod
    def BFreq(freq : np.ndarray, energy : np.ndarray) -> np.ndarray:

        nfreq = len(freq)
        wfreq = np.zeros((nfreq), dtype=np.complex128, order='F')

        for ifreq in range(nfreq):
            wfreq[ifreq] = 1.0 / (1j * freq[ifreq] - energy)

        return wfreq
    
    @staticmethod
    def BTau(tau : np.ndarray, energy : np.ndarray) -> np.ndarray:

        ntau = len(tau)
        wtau = np.zeros((ntau), dtype=np.complex128, order='F')

        pi = np.pi
        beta = tau[0]/(np.cos(pi*(ntau-0.5)/ntau) + 1.0)*2.0
        machep = np.finfo(np.float64).eps

        if (abs(energy) < 1.0e-12):
            print("Zero energy in Bare.BTau. impossible")
            exit()

        for itau in range(ntau):
            taumod = (tau[itau] % beta)
            unitnum = int(tau[itau] - taumod)/beta
            if (taumod < machep):
                unitnum = unitnum - 1
            taunew = tau[itau] - beta*unitnum

            if (energy > 0):
                wtau[itau] = -np.exp(-energy*beta, dtype=np.complex128) \
                    * (1.0 - 1.0/(np.exp(energy*beta, dtype=np.complex128) - 1))
            else:
                wtau[itau] = -np.exp(-energy*(taunew-beta), dtype=np.complex128) \
                    * (1.0/np.exp(energy*beta, dtype=np.complex128) - 1)
        return wtau


    @staticmethod
    def FLocFreq(freq : np.ndarray, hloc : np.ndarray)->np.ndarray:

        norb, _, ns = hloc.shape
        nfreq = len(freq)
        tempmat = np.zeros((norb, norb), dtype=np.complex128, order='F')
        tempmat2 = np.zeros((norb, norb), dtype=np.complex128, order='F')
        gloc = np.zeros((norb, norb, ns, nfreq), dtype=np.complex128, order='F')

        for js in range(ns):
            tempmat = hloc[:, :, js]
            w, v = Common.HermitianEigenCmplx(tempmat)
            gfreq = np.zeros((nfreq, norb), dtype=np.complex128, order='F')

            for iorb in range(norb):
                gfreq[:, iorb] = Bare.FFreq(freq, w[iorb])

            for ifreq in range(nfreq):
                for jorb in range(norb):
                    for iorb in range(norb):
                        tempmat2[iorb, jorb] = v[iorb, jorb]*gfreq[ifreq, jorb]
            
                gloc[:, :, js, ifreq] = tempmat2@(np.conjugate(v.T))

        return gloc
    
    @staticmethod
    def FLatFreq(freq : np.ndarray, hlatt : np.ndarray) -> np.ndarray:

        norb, _, ns, nk = hlatt.shape
        nfreq = len(freq)
        glatt = np.zeros((norb, norb, ns, nk, nfreq),dtype=np.complex128, order='F')

        for ik in range(nk):
            glatt[..., ik, :] = Bare.FLocFreq(freq, hlatt[..., ik])
        
        return glatt
    
    @staticmethod
    def FLocTau(tau : np.ndarray, hloc : np.ndarray) -> np.ndarray:
        
        norb, _, ns = hloc.shape
        ntau = len(tau)
        tempmat = np.zeros((norb, norb), dtype=np.complex128, order='F')
        tempmat2 = np.zeros((norb, norb), dtype=np.complex128, order='F')
        gloc = np.zeros((norb, norb, ns, ntau), dtype=np.complex128, order='F')

        for js in range(ns):
            tempmat = hloc[:, :, js]
            w, v = Common.HermitianEigenCmplx(tempmat)
            gtau = np.zeros((ntau, norb), dtype=np.complex128, order='F')

            for iorb in range(norb):
                gtau[:, iorb] = Bare.FTau(tau, w[iorb])
                
            for itau in range(ntau):
                for jorb in range(norb):
                    for iorb in range(norb):
                        tempmat2[iorb, jorb] = v[iorb, jorb]*gtau[itau, jorb]
                
                gloc[:, :, js, itau] = tempmat2@(np.conjugate(v.T))
        
        return gloc
    
    @staticmethod
    def FLatTau(tau : np.ndarray, hlatt : np.ndarray) -> np.ndarray:

        norb, _, ns, nk = hlatt.shape
        ntau = len(tau)
        glatt = np.zeros((norb, norb, ns, nk, ntau),dtype=np.complex128, order='F')

        for ik in range(nk):
            glatt[..., ik, :] = Bare.FLocTau(tau, hlatt[..., ik])

        return glatt

    @staticmethod
    def BLocFreq(freq : np.ndarray, hloc : np.ndarray) -> np.ndarray:

        norb, _, ns, _ = hloc.shape
        nfreq = len(freq)
        tempmat = np.zeros((norb, norb), dtype=np.complex128, order='F')
        tempmat2 = np.zeros((norb, norb), dtype=np.complex128, order='F')
        wloc = np.zeros((norb, norb, ns, ns, nfreq), dtype=np.complex128, order='F')

        for ks in range(ns):
            for js in range(ns):
                tempmat[:, :, js, ks] = hloc[:, :, js, ks]
                w, v = Common.HermitianEigenCmplx(tempmat)
                wfreq = np.zeros((nfreq, norb), dtype=np.complex128, order='F')

                for iorb in range(norb):
                    wfreq[:, iorb] = Bare.BFreq(freq, w[iorb])
                
                for ifreq in range(nfreq):
                    for jorb in range(norb):
                        for iorb in range(norb):
                            tempmat2[iorb, jorb] = v[iorb, jorb]*wfreq[ifreq, jorb]
                    
                    wloc[:, :, js, ks, ifreq] = tempmat2@(np.conjugate(v.T))
        
        return wloc
    
    @staticmethod
    def BLatFreq(freq : np.ndarray, hlatt : np.ndarray) -> np.ndarray:

        norb, _, ns, _, nk = hlatt.shape
        nfreq = len(freq)
        wlatt = np.zeros((norb, norb, ns, ns, nk, nfreq), dtype=np.complex128, order='F')

        for ik in range(nk):
            wlatt[..., ik, :] = Bare.BLocFreq(freq, hlatt[..., ik])

        return wlatt
    
    @staticmethod
    def BLocTau(tau : np.ndarray, hloc : np.ndarray) -> np.ndarray:

        norb, _, ns, _ = hloc.shape
        ntau = len(tau)
        tempmat = np.zeros((norb, norb), dtype=np.complex128, order='F')
        tempmat2 = np.zeros((norb, norb), dtype=np.complex128, order='F')
        wloc = np.zeros((norb, norb, ns, ns, ntau), dtype=np.complex128, order='F')

        for ks in range(ns):
            for js in range(ns):
                tempmat = hloc[:, :, js, ks]
                w, v = Common.HermitianEigenCmplx(tempmat)
                wtau = np.zeros((ntau, norb), dtype=np.complex128, order='F')

                for iorb in range(norb):
                    wtau[:, iorb] = Bare.BTau(tau, w[iorb])
                
                for itau in range(ntau):
                    for jorb in range(norb):
                        for iorb in range(norb):
                            tempmat2[iorb, jorb] = v[iorb, jorb]*wtau[itau, jorb]
                    
                    wloc[:, :, js, ks, itau] = tempmat2@(np.conjugate(v.T))
        
        return wloc
    
    @staticmethod
    def BLatTau(tau : np.ndarray, hlatt : np.ndarray) -> np.ndarray:

        norb, _, ns, _, nk = hlatt.shape
        ntau = len(tau)
        wlatt = np.zeros((norb, norb, ns, ns, nk, ntau), dtype=np.complex128, order='F')

        for ik in range(nk):
            wlatt[..., ik, :] = Bare.BLocTau(tau, hlatt[..., ik])

        return wlatt
    
